msgraph/core/graph_client_factory.py

```diff
@@ -0,0 +1,117 @@
+            base_url (NationalClouds): a supported Microsoft Graph cloud endpoint.
+            timeout (httpx.Timeout):Default connection and read timeout values for all session
+                                    requests.Specify a tuple in the form of httpx.Timeout(
+                                    REQUEST_TIMEOUT, connect=CONNECTION_TIMEOUT),
+            client (Optional[httpx.AsyncClient]): A custom AsynClient instance from the
+                                    python httpx library
+        """
+        self.api_version = api_version
+        self.base_url = base_url
+        self.timeout = timeout
+        self.client = client
+
+    def create_with_default_middleware(
+        self, token_provider: AccessTokenProvider
+    ) -> httpx.AsyncClient:
+        """Constructs native HTTP AsyncClient(httpx.AsyncClient) instances configured with
+        a custom transport loaded with a default pipeline of middleware.
+        Returns:
+            httpx.AsycClient: An instance of the AsyncClient object
+        """
+        if not self.client:
+            self.client = httpx.AsyncClient(
+                base_url=self._get_base_url(), timeout=self.timeout, http2=True
+            )
+        current_transport = self.client._transport
+        middleware = self._get_default_middleware(token_provider, current_transport)
+
+        self.client._transport = AsyncKiotaTransport(
+            transport=current_transport, middleware=middleware
+        )
+        return self.client
+
+    def create_with_custom_middleware(
+        self, middleware: Optional[List[BaseMiddleware]]
+    ) -> httpx.AsyncClient:
+        """Applies a custom middleware chain to the HTTP Client
+
+        Args:
+            middleware(List[BaseMiddleware]): Custom middleware list that will be used to create
+            a middleware pipeline. The middleware should be arranged in the order in which they will
+            modify the request.
+        """
+        if not self.client:
+            self.client = httpx.AsyncClient(
+                base_url=self._get_base_url(), timeout=self.timeout, http2=True
+            )
+        current_transport = self.client._transport
+
+        self.client._transport = AsyncKiotaTransport(
+            transport=current_transport, middleware=middleware
+        )
+        return self.client
+
+    def _get_base_url(self):
+        """Helper method to set the base url"""
+        base_url = self.base_url + '/' + self.api_version
+        return base_url
+
+    def _get_default_middleware(
+        self, token_provider: AccessTokenProvider, transport: httpx.AsyncBaseTransport
+    ) -> GraphMiddlewarePipeline:
+        """
+        Helper method that constructs a middleware_pipeline with the specified middleware
+        """
+        middleware_pipeline = GraphMiddlewarePipeline(transport)
+        middleware = [
+            GraphAuthorizationHandler(token_provider),
```

Comment:
```suggestion
```
not needed anymore since the request adapter does that for us on the abstract request information object now.

Reply:
This is specifically for users who intend to use the core as a stand alone package to make http requests.

Reply:
yep, talked offline about it, two factories will do the trick to cover the different scenarios.